### PR TITLE
[jvm-packages] Support Ranker

### PR DESCRIPTION
**Pull request by wbo4958**

### CODE REVIEW

**Comment by wbo4958** [comment]
Hi @trivialfis, @eordentlich Could you help review it? Thx

**Comment by eordentlich** [comment]
Would be good to explain how this resolves the issue raised in this comment and issue linked therein:
https://github.com/dmlc/xgboost/pull/10639#issuecomment-2328198404

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostRanker.scala [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+  /**
+   * Preprocess the dataset to meet the xgboost input requirement
+   *
+   * @param dataset
+   * @return
+   */
+  override private[spark] def preprocess(dataset: Dataset[_]): (Dataset[_], ColumnIndices) = {
+    val (output, columnIndices) = super.preprocess(dataset)
+    (output.sortWithinPartitions(getGroupCol), columnIndices)
```
[the remark this answers is not in the document]
Comment (eordentlich) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostRanker.scala:72]: Does this preprocess [even get called](https://github.com/dmlc/xgboost/blob/96bbf80457087260ea5bf4bce0b58bd6b53f6e3b/jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostEstimator.scala#L403-L408) if plugin is enabled?   If not, partition might not be sorted.

**Review by eordentlich — COMMENTED** [review]
Some questions re plugin and preprocess.

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostRanker.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (wbo4958) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostRanker.scala]: My bad. Fixed this issue. Please help review it again. Thx very much.

**Comment by wbo4958** [comment]
Hi @eordentlich @trivialfis, Could you help take a look at it. Thx very much.

**Review by eordentlich — COMMENTED** [review]
This resolves the issue with the plugin and sorted partitions (and nice to see the test for this case too), but still wondering how that partition sort is computed by the spark-rapids plugin when enabled.  Is done on the GPU?

Also, does this PR resolve the issue I reference in an earlier comment?

@trivialfis should take a look as well.

**Review by eordentlich — APPROVED** [review]
👍

**Comment by trivialfis** [comment]
Will look into this tomorrow.

**Comment by trivialfis** [comment]
> Also, does this PR resolve the issue I reference in an earlier comment?

Thank you for raising that. I will be looking into it along with other LTR issues/feature requests after sorting out some of the work on external memory.

**Review by trivialfis — APPROVED** [review]
Looks good to me, assuming all tests can pass.

**Pull request merged**